pandas/_testing/asserters.py

```diff
@@ -317,8 +320,8 @@ def _check_types(left, right, obj="Index") -> None:
 
         # Skip exact dtype checking when `check_categorical` is False
         if check_categorical:
-            assert_attr_equal("dtype", left, right, obj=obj)
             if is_categorical_dtype(left.dtype) and is_categorical_dtype(right.dtype):
+                assert_attr_equal("dtype", left, right, obj=obj)
```

Comment:
why is this moved?

Reply:
If this is above, the dtype check is also performed for non-categorical data.

Concretely, `assert_index_equal(NumericIndex[int32], NumericIndex[int64], exact="equiv")` should return pass, but raises anAssertionError without this change.

Reply:
I've refactored this in #41980, so this file isn't touched in this PR any more.

Reply:
I've refactored this in #41980, so this file isn't touched in this PR any more.